### PR TITLE
xtensa-build-zephyr.py: don't leave the install of an old build behind

### PR DESCRIPTION
**Pull request by marc-hb**
In theory a failure from this script should immediately stop the show and no one should then use the staging directory.

In practice we cannot make sure, especially not with shell scripts. So make sure no previous build of a requested platform is left behind.

Signed-off-by: Marc Herbert <marc.herbert@intel.com>

### CODE REVIEW

File: scripts/xtensa-build-zephyr.py [hunk 1]
```diff
@@ -477,6 +477,26 @@ def get_build_and_sof_version(abs_build_dir):
...
+	rmtree_if_exists(STAGING_DIR / "sof-info" / platform)
+
+	sof_output_dir = STAGING_DIR / "sof"
+
+	# --use-platform-subdir
+	rmtree_if_exists(sof_output_dir / platform)
+
+	# Remaining .ri and .ldc files
+	for f in sof_output_dir.glob(f"**/sof-{platform}.*"):
+		os.remove(f)
```
Comment (kv2019i) [inline scripts/xtensa-build-zephyr.py:497]: Wouldn't this delete all platforms, even those not build in this run? Or that is the intention?

**Review by kv2019i — COMMENTED** [review]
This seems pretty scary. I'm guessing you've double-checked and tripled checked this can in no circumstances go haywire.

File: scripts/xtensa-build-zephyr.py (hunk 1 above)
Comment (marc-hb) [reply scripts/xtensa-build-zephyr.py:497]: It starts by cleaning only the platforms _requested_ in this run, see the `{platform}` argument.

**Comment by marc-hb** [comment]
Unrelated crash in https://sof-ci.01.org/sofpr/PR6819/build2937/devicetest/index.html

**Comment by marc-hb** [comment]
SOFCI TEST

**Review by aborisovich — COMMENTED** [review]
I'll test this on Windows and let you know asap.

**Review by miRoox — APPROVED** [review]

**Review by aborisovich — APPROVED** [review]
Windows CI works, change works as intended.
Performed following test:
1. execute `python sof\scripts\xtensa-build-zephyr.py tgl -i IPC4`
2. execute `python sof\scripts\xtensa-build-zephyr.py mtl`
3. execute `python sof\scripts\xtensa-build-zephyr.py mtl`

Result:
After step 2 and step 3 TGL was build and output was not removed.
After step 3 MTL binaries from build-sof-staging have updated timestamps (removed binaries from step 2).
PR is good to go.

Still I'll rumble a bit again on our pythons, just FYI as a future refactor material:
1. Copying FW binaries and tools to `build-sof-staging` should be done as [CMake Install](https://cmake.org/cmake/help/latest/command/install.html) step.
2. Then, what @marc-hb done in this PR should be done using
[CMake option](https://cmake.org/cmake/help/v3.1/manual/cmake.1.html#options) --clean-first.
This flag was introduced quite recently but it does exactly what @marc-hb did here.
It removes files installed as part of install step, removes objects and binaries created by build, and most importantly, does not delete toolchain check discovery files so you basically can easily rebuild skipping this process:
image

**Comment by marc-hb** [comment]
Thanks a lot for the review and testing.

> Copying FW binaries and tools to build-sof-staging should be done as [CMake Install](https://cmake.org/cmake/help/latest/command/install.html) step.

That seems more logical but:
- This /lib/firmware/ "installation" looks nothing like installing regular C/C++ software. I'm afraid this would require twisting CMake's arm a lot. CMake's "install" documentation is full of concepts totally irrelevant here like headers, static/shared libraries, executables,... There is a "custom installation logic" that can delegate to an external (Python?!) script but then everything is hidden from CMake and do features like `--clean-first` still work?
- It's hard enough to find people willing to make any change to build and test scripts; at least Python scripts are less scary and much more popular than CMake scripts. In fact CMake's new and main competitor "meson" adopted a Python-looking syntax.
- We will always need this script to at least iterate on multiple platforms. I hope it can become smaller (the XCC versions should especially be moved) but it will never be zero.

**Pull request merged**